Valigator.Core/Data.cs

```diff
@@ -117,5 +119,37 @@ private static IDataContainer<TValue> GetErrorDataContainer(IDataContainer<TValu
 
 		public static implicit operator TValue(Data<TValue> data)
 			=> data.Value;
+
+		public override bool Equals(object obj) 
+			=> obj is Data<TValue> data 
+				&& this == data;//if (obj is Data<TValue> data)//	return this == data;//else//	return false;
+
+		public static bool operator ==(Data<TValue> x, Data<TValue> y) 
+			=> x.DataDescriptor.Equals(y.DataDescriptor)
+				&& (x._value.Equals(y._value) 
+					|| (typeof(TValue).IsArray && StructuralComparisons.StructuralEqualityComparer.Equals(x._value, y._value)));	
+
+		public static bool operator !=(Data<TValue> x, Data<TValue> y)
+			=> !x._value.Equals(y._value);
+
+		public override int GetHashCode()
+		{
+			int hashCode = 943777100;
+			var typeParameter = typeof(TValue);
+			if (typeParameter.IsArray || typeParameter.IsEnum)
```

Comment:
I'm not sure `IsEnum` is right here. I think you might be looking for something like this that checks if it is IEnumerable:
`return type.GetInterfaces().Any(t => t.GetGenericTypeDefinition() == typeof(IEnumerable<>));`

Reply:
I think that'll do the same thing, it just doesn't use reflection, right?

Reply:
I put this change in, but I think it does the same thing as IsEnum.

Reply:
`typeof(int[]).IsEnum` returns false: https://docs.microsoft.com/en-us/dotnet/api/system.type.isenum?view=net-5.0

Reply:
That's where the `IsArray || IsEnum` would come in.

Reply:
What is the goal of this line? I interpreted it as intending to say "If it's an array or an IEnumerable" then true. `IsEnum` tells if it's an enum type (ie: `public enum Colors { Red, Blue, Green }`). For this reason, `typeof(IEnumerable<int>).IsEnum` is also false, but `typeof(Colors).IsEnum` is true.

Reply:
So, the IEnumerable interface for arrays is implemented at runtime - not in the Array class declaration (hence why there's an IsArray And IsEnumerable on a type).
I found that the ISEnum is a couple milliseconds faster on average than the linq interface match, so I'll switch it back the that

Reply:
The `IsArray` isn't what I'm contesting here. I'm saying that `IsEnum` doesn't check if a type is an `IEnumerable`.

Reply:
Right.  That was my mistake.  I'll run some tests for speed and try  `typeParameter is IEnumerable` vs the linq.

Reply:
`typeParameter is IEnumerable` is slightly faster and is actually checking if it's an IEnumerable.

Reply:
Actually, I think you're right checking the interfaces is the best option here.